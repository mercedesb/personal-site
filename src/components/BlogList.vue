<template>
  <div class='PageContent PageContent--fullWidth'>
    <div v-if="loading">
      <BlogListItem v-for="n in 10" :loading="true" :key="n"></BlogListItem>
    </div>
    <div v-if="readyToDisplay">
      <BlogListItem v-if="featuredPost"
        v-bind="featuredPost"
        :key="featuredPost.id"
        :color="color"
      ></BlogListItem>
      <BlogListItem v-for="blogPost in remainingPosts"
        v-bind="blogPost"
        :key="blogPost.id"
        :color="color"
      ></BlogListItem>
    </div>
  </div>
</template>

<script>
import BlogListItem from './BlogListItem.vue'

export default {
  components: {
    BlogListItem
  },
  props: {
    color: String,
    posts: Array,
    isFirstPage: Boolean,
    loading: Boolean
  },
  computed: {
    readyToDisplay() {
      return !this.loading && this.posts && this.posts.length
    },
    featuredPost () {
      return {
        featured: this.isFirstPage,
        ...this.posts[0]
      }
    },
    remainingPosts () {
      return this.posts.slice(1)
    }
  }
}
</script>
