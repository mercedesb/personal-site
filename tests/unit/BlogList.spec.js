import BlogList from "@/components/BlogList.vue";
import { TestUtility } from "./TestUtility";

describe("BlogList", () => {
  let component;

  const initialProps = {
    color: "gray"
  };

  const shallow = propsData =>
    TestUtility.shallow(BlogList, {
      propsData: {
        ...initialProps,
        ...propsData
      }
    });

  describe("Snapshots", () => {
    describe("with a color passed in", () => {
      it("matches snapshot", () => {
        component = shallow({
          posts: TestUtility.blogPosts,
          isFirstPage: true
        });
        expect(component.element).toMatchSnapshot();
      });
    });

    describe("without a color passed in", () => {
      it("matches snapshot", () => {
        component = shallow({
          color: undefined,
          posts: TestUtility.blogPosts,
          isFirstPage: true
        });
        expect(component.element).toMatchSnapshot();
      });
    });

    describe("without posts", () => {
      it("matches snapshot", () => {
        component = shallow({
          posts: [],
          isFirstPage: true
        });
        expect(component.element).toMatchSnapshot();
      });
    });
  });

  describe("Computed", () => {
    describe("featuredPost", () => {
      it("has featuredPost populated", () => {
        component = shallow({
          posts: TestUtility.blogPosts,
          isFirstPage: true
        });
        expect(component.vm.featuredPost).toBeDefined();
      });

      it("returns the first post of the first page as featured", () => {
        component = shallow({
          posts: TestUtility.blogPosts,
          isFirstPage: true
        });
        const expected = {
          featured: true,
          ...component.vm.posts[0]
        };
        expect(component.vm.featuredPost).toEqual(expected);
      });

      it("returns the first post of not the first page as not featured", () => {
        component = shallow({
          posts: TestUtility.blogPosts,
          isFirstPage: false
        });
        const expected = {
          featured: false,
          ...component.vm.posts[0]
        };
        expect(component.vm.featuredPost).toEqual(expected);
      });
    });

    describe("remainingPosts", () => {
      beforeEach(() => {
        component = shallow({
          posts: TestUtility.blogPosts,
          isFirstPage: true
        });
      });

      it("has remainingPosts populated", () => {
        expect(component.vm.remainingPosts).toBeDefined();
      });

      it("returns the second through end post", () => {
        const expected = component.vm.posts.slice(1);
        expect(component.vm.remainingPosts).toEqual(expected);
      });
    });
  });
});
