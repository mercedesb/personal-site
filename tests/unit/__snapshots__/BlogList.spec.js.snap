// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlogList Snapshots when loading matches snapshot 1`] = `
<div
  class="PageContent PageContent--fullWidth"
>
  <!---->
   
  <!---->
</div>
`;

exports[`BlogList Snapshots with a color passed in matches snapshot 1`] = `
<div
  class="PageContent PageContent--fullWidth"
>
  <!---->
   
  <div>
    <bloglistitem-stub
      color="gray"
      featured="true"
      id="blogPost1"
      maincontent=""
      preamble="blog post 1 preamble"
      publishdate="2018-04-25T00:00-05:00"
      tags="Test"
      title="blog post 1 title"
      urlsegment="/blogPost1"
    />
     
    <bloglistitem-stub
      color="gray"
      id="blogPost2"
      maincontent=""
      preamble="blog post 2 preamble"
      publishdate="2018-04-23T00:00-05:00"
      tags=""
      title="blog post 2 title"
      urlsegment="/blogPost2"
    />
    <bloglistitem-stub
      color="gray"
      id="blogPost3"
      maincontent=""
      preamble="blog post 3 preamble"
      publishdate="2018-04-22T00:00-05:00"
      tags=""
      title="blog post 3 title"
      urlsegment="/blogPost3"
    />
  </div>
</div>
`;

exports[`BlogList Snapshots without a color passed in matches snapshot 1`] = `
<div
  class="PageContent PageContent--fullWidth"
>
  <!---->
   
  <div>
    <bloglistitem-stub
      color="loading-gray"
      featured="true"
      id="blogPost1"
      maincontent=""
      preamble="blog post 1 preamble"
      publishdate="2018-04-25T00:00-05:00"
      tags="Test"
      title="blog post 1 title"
      urlsegment="/blogPost1"
    />
     
    <bloglistitem-stub
      color="loading-gray"
      id="blogPost2"
      maincontent=""
      preamble="blog post 2 preamble"
      publishdate="2018-04-23T00:00-05:00"
      tags=""
      title="blog post 2 title"
      urlsegment="/blogPost2"
    />
    <bloglistitem-stub
      color="loading-gray"
      id="blogPost3"
      maincontent=""
      preamble="blog post 3 preamble"
      publishdate="2018-04-22T00:00-05:00"
      tags=""
      title="blog post 3 title"
      urlsegment="/blogPost3"
    />
  </div>
</div>
`;

exports[`BlogList Snapshots without posts matches snapshot 1`] = `
<div
  class="PageContent PageContent--fullWidth"
>
  <div>
    <bloglistitem-stub
      color="loading-gray"
      loading="true"
      maincontent=""
      urlsegment="/"
    />
    <bloglistitem-stub
      color="loading-gray"
      loading="true"
      maincontent=""
      urlsegment="/"
    />
    <bloglistitem-stub
      color="loading-gray"
      loading="true"
      maincontent=""
      urlsegment="/"
    />
    <bloglistitem-stub
      color="loading-gray"
      loading="true"
      maincontent=""
      urlsegment="/"
    />
    <bloglistitem-stub
      color="loading-gray"
      loading="true"
      maincontent=""
      urlsegment="/"
    />
    <bloglistitem-stub
      color="loading-gray"
      loading="true"
      maincontent=""
      urlsegment="/"
    />
    <bloglistitem-stub
      color="loading-gray"
      loading="true"
      maincontent=""
      urlsegment="/"
    />
    <bloglistitem-stub
      color="loading-gray"
      loading="true"
      maincontent=""
      urlsegment="/"
    />
    <bloglistitem-stub
      color="loading-gray"
      loading="true"
      maincontent=""
      urlsegment="/"
    />
    <bloglistitem-stub
      color="loading-gray"
      loading="true"
      maincontent=""
      urlsegment="/"
    />
  </div>
   
  <!---->
</div>
`;
